src/machine/system_calls.rs

```diff
@@ -1328,11 +1333,18 @@ impl Machine {
                     }
                 )
             }
-            None => Ok(if let Some(load_context) = self.load_contexts.last() {
-                load_context.module
-            } else {
-                atom!("user")
-            }),
+            None => Ok(self.default_module_name()),
+        }
+    }
+
+    /// Returns the default module name: the last [`module`](super::LoadContext::module)
+    /// in [`load_contexts`](Self::load_contexts), or `"user"` if there are none.
+    #[inline]
+    pub(crate) fn default_module_name(&self) -> Atom {
```

Comment:
Would it be possible and sensible (and acceptably efficient) to move the definition of `default_module/1` *entirely* to Prolog? It is generally an advantage to move definitions to Prolog, also since it adds greater&nbsp;motivation for making the underlying&nbsp;engine more efficient, which in turn benefits also other Prolog&nbsp;parts.

Reply:
I'm not sure how to efficiently call prolog predicates from rust code, especially given that this sits in the hot path of `fast_call`

Reply:
Thank, it all already looks terrific as is! Maybe we can improve this specific&nbsp;aspect in a much later stage of development, for instance by automatically JIT-compiling certain Prolog&nbsp;definitions during&nbsp;bootstrapping.

Reply:
I feel like that would still be a bad idea for anything that `fast_call` relies on, since if we ever dare to call another predicate from within this prolog implementation, it could then infinitely recurse.

This reminds me of instances where people wanted to call C code from within the bootup sequence (that needs to be written in assembly) on embedded devices. As soon as that C code dares to refer to global variables (which were then still unitialized), things would break, badly.

Reply:
The idea would be to automatically generate the Rust or assembly code that you now had to write manually, from a Prolog-based definition of this predicate (a single&nbsp;fact in this case); we could do this already now, with an existing Scryer Prolog binary.